### PR TITLE
Fix plugin API in globally installed Prettier and introduce optional --plugin-search-dir

### LINKED ISSUE #4000: Plugin API not working with globally installed modules
Currently, globally installed plugins don't seem to be discovered by the CLI:
```bash
npm install -g prettier @prettier/plugin-php
prettier --plugin=@prettier/plugin-php --parser=php test.php
# Error: Cannot find module '@prettier/plugin-php'

npm init --yes
npm install prettier @prettier/plugin-php
prettier --plugin=@prettier/plugin-php --parser=php test.php
# works
```
It seems that `read-pkg-up` is just looking for the closest `package.json` file, which doesn't work for global installs 
https://github.com/prettier/prettier/blob/1b61b819d7f58c645c97a89b26e86915cad3c1f1/src/common/load-plugins.js#L4

(in case you're wondering why i'm explicitly passing the plugin: There also seems to be an issue with automatic discovery, running the command without results in `SyntaxError: Unexpected token` (see https://github.com/prettier/plugin-php/issues/55))

### PR DESCRIPTION
**Pull request by kachkaev**
Closes #4000 🎉 

Instead of searching for `dependencies` and `devDependnencies`, in the nearest `package.json`, Prettier now searches for `["prettier-plugin-*/package.json", "@prettier/plugin-*/package.json"]` in the nearest parent `node_modules` directory. This makes it possible for a globally installed instance to find plugins despite the lack of `package.json` up the tree. What is also good about this approach is that plugins are discoverable when they are a part of some boilerplate packages like `create-react-app`. Their philosophy is to stuff a project with all bells and whistles at a cost of just a single dependency in `package.json`.

Rather than matching directories like `prettier-plugin-*` and `@prettier/plugin-*` inside `node_modules`, `loadPlugins()` is looking for `prettier-plugin-*/package.json` and `@prettier/plugin-*/package.json`. I decided to do this for two reasons::

* directories alone cannot be searched by `globby`, which has already been in Prettier dependencies (so no need to reinvent the wheel)
* checking the presence of `package.json` file is more robust – there'll be no false-positives on empty dirs

This change in Prettier logic has been [approved](https://github.com/prettier/prettier/issues/4000#issuecomment-374574693) by @azz, the author of #3536 and #3624.

---

__UPD:__ After a few iterations on this PR, I also added an _optional_ `--plugin-search-dir` CLI parameter, justification of which you will find below.

### CODE REVIEW

**Comment by suchipi** [comment]
Does this only look in the first node_modules dir upwards? Why not go up forever? Yarn workspaces and differences between npm and yarn are some things to consider here.

**Comment by kachkaev** [comment]
Do you think we should search for all `node_modules` up the tree? To be honest, I doubt the usefulness of this. No matter if we use `npm` or `yarn`, project's main prettier (out of several  duplicates, if any) will be in `path/to/project/node_modules/prettier`. There'll be also `path/to/project/node_modules/.bin/prettier`, which links to `../lib/node_modules/prettier/bin-prettier.js`. Given these locations, plugins should work exactly the same way as it did for the previous implementation, not more not less.

There is one extra bit that I've added a bit later – see https://github.com/prettier/prettier/pull/4192/commits/61d87afd98edd9e9eff2a59baab75de6314847e8. A directory with node modules is searched in two steps now. The first one succeeds for standard prettier (including both local and global ones), while the second step is useful for the `npm link`-ed version of prettier and in tests.

**Comment by kachkaev** [comment]
I fixed one typo and the tests started to fail. Looks like we need a better algorithm to decide how to find parent `node_modules`, in which we'll search for package directories 🤔 

`findDirUp(__dirname, "node_modules")`

works for
* local prettier instances, any `cwd`
* global prettier instances, any `cwd`

does not work for
* local prettier instances installed using `npm link`
* tests

`findDirUp(process.cwd(), "node_modules")`

works for
* local prettier instances, assuming `cwd` is within the project
* tests

does not work for
* local prettier instances if `cwd` is outside project dir
* global prettier instances

**Comment by kachkaev** [comment]
Perhaps, if we add some special `if` for the testing environment, that'd be it, however, this option smells. In any case, we should not trust cwd as it used to be because of https://github.com/prettier/prettier/issues/4000#issuecomment-374309824

**Comment by kachkaev** [comment]
```js
  const nodeModulesDir =
    findDirUp(process.env.PRETTIER_PLUGINS_ROOT, "node_modules") ||
    findDirUp(__dirname, "node_modules") ||
    findDirUp(process.cwd(), "node_modules");
```
🤔

**Comment by kachkaev** [comment]
Or how about `--plugin-location` that would point to node_modules? It will be needed only need to be set in two edge cases
* in tests
* when prettier is installed using `npm link`

otherwise `findDirUp(__dirname, "node_modules")` is used and everyone is happy. No more confusing cwd.

**Comment by kachkaev** [comment]
This PR is updated with a new `--plugin-search-dir` option. When it is not defined, prettier searches for plugins in the nearest `node_nodules` above it. Otherwise, the plugins are searched in `node_modules` subdirectory of the given path. This makes plugins potentially auto-discoverable in all these cases:
* local prettier
* global prettier
* tests (with `--plugin-search-dir` provided)
* `npm-linke`d instance of prettier (with `--plugin-search-dir` provided)

There is one tests that is still failing – it is to do with getting help for a specific prettier option (`prettier --help tab-width --parser=bar --plugin-search-dir=.` – I can't manage to pass the value of `--plugin-search-dir=.` into `getSupportInfo()` 😢 

@azz could you please help me with passing one remaining test?

**Comment by suchipi** [comment]
Can you really call plugins "auto"-discoverable if you have to specify an option pointing to where they are? 😅 

As an aside, `require("module").globalPaths` returns the global module paths where npm installs stuff, but not the yarn global path (which you can get via `yarn global dir`). 

Could we combine pkg-search-up with `require("module").globalPaths` and `yarn global dir` to get what we want?

**Comment by kachkaev** [comment]
Yeah, I also see `"auto"-discoverable` a slightly funny term when we define `--plugin-search-dir`, but  even in this case there still some part of auto-discoverability: the plugins do get searched amongst `node_modules` 😄 

> Could we combine pkg-search-up with require("module").globalPaths and yarn global dir to get what we want?

I'm afraid that's not the way to go, because we don't want ambiguity caused by multiple plugin installs on one machine. Should a local instance of prettier fall back to global yarn or npm plugins if none are found locally? And what about if some are found? Guess the answer is now for both cases, unless there is guidance from a user (i.e. a custom-defined `--plugin-search-dir`). Also how do we make `npm link`-ed prettier search for plugins in the current project's directory [without](https://github.com/prettier/prettier/issues/4000#issuecomment-374309824) using `process.cwd` and `--plugin-search-dir`? There seems to be now way to achieve that. Turns out, `--plugin-search-dir` is the least of evils that covers all possible use cases for plugin search.

In future, we can support multiple `--plugin-search-dir`s just as we support multiple `--plugin` options now.

**Comment by suchipi** [comment]
I think I misunderstood the goals and constraints of the PR. Disregard my comments until I have more time to look into this properly

**Comment by kachkaev** [comment]
You probably thought that `--plugin-search-dir` is compulsory – and it's my fault because I forgot to mention that it's optional in the first comment. Fixed that now.

**Comment by kachkaev** [comment]
Spotted a bug in `--plugin-search-dir` autodetection when prettier is installed with yarn.

When there exists `/path/to/project/node_modules/prettier/node_modules` folder, prettier incorrectly starts searching for plugins in it rather than in `/path/to/project/node_modules`. Guess we fix that by replacing 

```
findDirUp(__dirname, "node_modules")
```

with

```
findDirUp(path.resolve(findDirUp(__dirname, "prettier"), ".."), "node_modules")
```

**Comment by azz** [comment]
Not a huge fan of the `--plugin-search-dir` option, it seems to add a lot of complexity. The original implementation was intentionally minimal.

I'd be interested to see if the [`cwd` approach](https://github.com/prettier/prettier/issues/4000#issuecomment-366613619) fixes the issue raised (passing explicit `--plugin=x` with a globally installed `x`), I think I would prefer that solution instead.

**Comment by kachkaev** [comment]
Hi @azz

I thought about setting `readPkgUp`'s `cwd` to Prettier's own directory, but declined this simple option. It would not allow for auto-search in the global prettier module, which I think is crucial and #4000 is also exactly about it. One should be able to do something like:

```bash
npm install --global prettier @prettier/plugin-a prettier-plugin-b prettier-plugin-c
```

And then enjoy

```bash
prettier "**/*.{js,md,a,b,c}"
```

in any folder on their machine. I know that having local dependencies is better, but despite that a global instance of prettier powered by several plugins can be useful, say, in editor extensions such as `prettier-atom`. Given this design constraint, we can't use `cwd` and we can's simply `readPkgUp`, because global npm packages don't have `package.json`.

Thus, we are left with searching for actual packages above the tree from where Prettier is and, to be honest, this is not way more complex:

```js
nodeModulesDir = findDirUp(
    path.resolve(findDirUp(__dirname, "prettier"), ".."),
    "node_modules"
);
const autoPluginNames = findPluginsInNodeModules(nodeModulesDir);
```
(we first find parent `prettier` dir and then `node_modules` above it)

Some complexity comes into play indeed, but only if we want to override this behaviour in two very rare edge cases:
* in tests (unless we copy prettier into `plugins/automatic/node_modules/prettier`)
* when prettier is installed with `npm link` and is being debugged

_This is only needed by advanced users, mostly while debugging stuff_, Everyone else simply enjoys autoloaded plugins no matter if install is global and what current `cwd` is. Option `--plugin-search-dir` is from the same boat as `--loglevel` and `--support-info`.

I'm open to other designs and I don't insist that my implementation needs to be merged into prettier. However, after several days of thinking on this problem I can't think of any other solutions that would satisfy all the constraints that the problem of plugin autodiscovery has.

**Comment by kachkaev** [comment]
Technically we can remove `--plugin-search-dir` and keep the tests passing. That will involve copying  prettier into `plugins/automatic/node_modules/prettier` before running `automatically loads '@prettier/plugin-*' from --plugin-search-dir"` and a couple of other tests. However, `npm link` case will be no longer covered and it won't be long till someone opens an issue titled _Prettier does not autoload plugins when installed using npm link_ 😃

**Comment by kachkaev** [comment]
@suchipi when do you think you could look at this? Curious to hear your opinion!

**Comment by suchipi** [comment]
@kachkaev I haven't had a chance to look at this yet but I haven't forgotten about it. I found somewhere that you linked to https://github.com/prettier/prettier-atom/issues/395 which helps me understand the context of and motivation for this change better. 

My main concerns are:
* Plugin detection should work for the use-cases you need in prettier-atom
* Plugin detection should work in monorepos (traverse upwards through multiple levels of local node_modules recursively)

Some "nice-to-have"s are that these would both work:
```sh
yarn global add prettier
yarn global add @prettier/plugin-php
prettier --write './**/*.php'
```
```sh
npm i -g prettier
npm i -g @prettier/plugin-php
prettier --write './**/*.php'
```

My kneejerk reaction to `--plugin-search-dir` is that I don't want to have to have it, but I don't fully understand why it's needed yet and I need to dig into this in order to understand.

I'll try to give this a thorough review sometime soon, but feel free to merge this without a review from me if it's ready.

**Comment by kachkaev** [comment]
I squashed all commits in this PR and rebased them on top of the latest master. Tests seem to pass. Original unsquashed commits are now in [`fix-global-plugin-api-before-rebase`](https://github.com/kachkaev/prettier/tree/fix-global-plugin-api-before-rebase) branch of my fork just in case.

@suchipi when I declare that `pluginSearchDir` is `since: "1.13.0"`, this option is not mentioned in some jests snapshots. What's the correct way of defining `since`?

It'd be great if someone could look at this PR at some point – we need to fix plugins in a global prettier setup 💯

**Comment by kachkaev** [comment]
Is there still interest in this PR? I'm really keen to see plugins working in a global instance of Prettier. Please help me to help! 🙂

**Comment by suchipi** [comment]
Hi @kachkaev, I have had this PR on my mind and I still have interest in it but haven't looked at it in a while because I've been pretty busy. I cannot promise when I will be able to look at this but I want prettier plugins to work with global installations, too.

**Comment by suchipi** [comment]
I apologize that this PR has been open for so long 🙏

**Comment by suchipi** [comment]
Also just saw your question about `since`, I don't know the answer but @ikatyang or @duailibe might

**Comment by ikatyang** [comment]
> when I declare that pluginSearchDir is since: "1.13.0", this option is not mentioned in some jests snapshots. What's the correct way of defining since?

I guess you meant `support-info.js.snap`, it only gets updated when we bump to a new version (i.e. v1.13) since we don't want to expose unreleased options for editor integration.

File: src/common/support.js [hunk 1]
```diff
@@ -192,7 +203,7 @@ const supportOptions = {
   }
 };
 
-function getSupportInfo(version, opts) {
+function getSupportInfo(version, opts, pluginSearchDir) {
```
[the remark this answers is not in the document]
Comment (ikatyang) [reply src/common/support.js]: Actually, `getSupportInfo` should be `fs`-unrelated, plugins should be loaded somewhere before passing in this function.

Ref: #4364

File: src/cli/util.js [hunk 2]
```diff
@@ -828,12 +828,17 @@ function initContext(context) {
...
-    plugins
-  }).options;
+  const supportOptions = getSupportInfo(
+    null,
+    {
+      showDeprecated: true,
+      showUnreleased: true,
+      showInternal: true,
+      plugins: plugins || [],
+      pluginsLoaded: typeof context.argv === "undefined"
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply src/cli/util.js]: Thanks for your review @ikatyang! Yeah, I've also noticed that #4364 requires some refactoring of this PR. I'm working on this now.

**Comment by kachkaev** [comment]
I guess `--plugin-search-dir` should be an array, just like `--plugin`. That is to allow autoloading from multiple folders, which may be useful in editor extensions. For instance, Prettier in Atom may want to autoload both local project's plugins as well as the global (editor-wide) ones.

I'll wait for https://github.com/prettier/prettier/pull/4341 to land to master before rebasing this PR. That's because `prettier.getFileInfo()` will have to accept `pluginSearchDirs` and I'd like to add that to tests.

**Comment by kachkaev** [comment]
`--plugin-search-dir` is an array now and https://github.com/prettier/prettier/pull/4364 is taken into account. I'm waiting for https://github.com/prettier/prettier/pull/4341 to be merged to add one more test.

@ikatyang I'm using [`find-parent-dir`](https://www.npmjs.com/package/find-parent-dir) instead  of `find-up` that you've suggested as its API slightly better fits what we need. This module is mocked in `tests_integrarion/__tests__/plugin-resolution.js` to test autoloading when `--plugin-search-dir` is not provided. I'm not sure how we can test autoloading without mocking given that Prettier's entry point within the repo has different parents than when it's installed in the wild. Apart from a need to partially mock autoloading, `load-plugins.js` should be covered pretty well:

> screen shot 2018-05-08 at 10 13 33

(I also built `prettier` and manually installed it as a global npm dependency – plugin resolution worked)

Please let me know if you have any questions about autoloading, it's not the most straightforward thing to figure out.

File: src/common/load-plugins.js [hunk 3]
```diff
@@ -16,45 +36,68 @@ function loadPlugins(plugins) {
...
-      )
-    )
-    .map(plugin => {
-      if (typeof plugin !== "string") {
-        return plugin;
-      }
-
-      const pluginPath = resolve.sync(plugin, { basedir: process.cwd() });
-      return Object.assign({ name: plugin }, eval("require")(pluginPath));
+  const externalPluginInfos = [];
```
Comment (ikatyang) [inline src/common/load-plugins.js]: It'd be better to not mutate array if possible as it's clearer, I'd prefer something like this:

```js
const externalManualLoadPluginInfos = plugins.map(pluginName => ({
  name: pluginName,
  requirePath: resolve.sync(pluginName, { basedir: process.cwd() })
}));

const externalAutoLoadPluginInfos = pluginSearchDirs
  .map(pluginSearchDir => {
    const resolvedPluginSearchDir = path.resolve(
      process.cwd(),
      pluginSearchDir
    );

    if (!isDirectory(pluginSearchDir)) {
      throw new Error(
        `${pluginSearchDir} does not exist or is not a directory`
      );
    }

    const nodeModulesDir = path.resolve(
      resolvedPluginSearchDir,
      "node_modules"
    );

    return findPluginsInNodeModules(nodeModulesDir).map(pluginName => ({
      name: pluginName,
      requirePath: resolve.sync(pluginName, {
        basedir: resolvedPluginSearchDir
      })
    }));
  })
  .reduce((a, b) => a.concat(b), []);

const externalPlugins = uniqBy(
  externalManualLoadPluginInfos.concat(externalAutoLoadPluginInfos),
  "requirePath"
).map(externalPluginInfo =>
  Object.assign(
    { name: externalPluginInfo.name },
    eval("require")(externalPluginInfo.requirePath)
  )
);
```

File: tests_integration/__tests__/plugin-resolution.js [hunk 4]
```diff
@@ -1,22 +1,108 @@
 "use strict";
```
Comment (ikatyang) [inline tests_integration/__tests__/plugin-resolution.js]: It seems the current implementation does not support `pluginSearchDirs` in config files, which we should do something like the following in `resolve-config.js`:

```js
options.pluginSearchDirs = options.pluginSearchDirs.map(
  pluginSearchDir => path.resolve(
    path.dirname(configFilePath), 
    pluginSearchDir
  )
)
```

**Comment by kachkaev** [comment]
Many thanks for your review comments @ikatyang, especially for a trick with mocking! It took me ages to figure out what how to test `findParentDir` and I still did it wrong in the end 😅 

One thing I did not address is supporting `pluginSearchDirs` in Prettier config. That's a great idea IMO, however, don't you think that `plugins` should be configurable this way too? If yes, will these two config file options supplement CLI/API options or will the latter replace the values in the config file?

How about starting a separate discussion in a new PR and too keep this one concise?

File: src/main/options.js [hunk 5]
```diff
@@ -20,6 +20,7 @@ function normalize(options, opts) {
 
   const supportOptions = getSupportInfo(null, {
     plugins: options.plugins,
+    pluginSearchDirs: options.pluginSearchDirs,
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply src/main/options.js]: Indeed – removing `pluginSearchDirs: options.pluginSearchDirs` in both cases did not break local tests. This must be an artefact of my earlier attempts prior to #4364.

**Comment by ikatyang** [comment]
> How about starting a separate discussion in a new PR and too keep this one concise?

Sure. And currently the options from `CLI`/`config file` will be overridden based on [`--config-precedence`](https://prettier.io/docs/en/cli.html#config-precedence), if we'd like to support supplement, we just need to `concat` them I guess.

File: tests_integration/runPrettier.js [hunk 6]
```diff
@@ -71,9 +71,6 @@ function runPrettier(dir, args, options) {
   // We cannot use `jest.setMock("get-stream", impl)` here, because in the
   // production build everything is bundled into one file so there is no
   // "get-stream" module to mock.
-  jest.spyOn(require(thirdParty), "getStream").mockImplementation(() => ({
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply tests_integration/runPrettier.js]: The reason for this change was that I wanted to sort third-party scripts alphabetically. I reverted that in https://github.com/prettier/prettier/pull/4192/commits/bfbc580af2f15851aff75b59491e03c259ccc718, but I'd be keen to delete it and thus to keep sorting. This will help read and write other third-party scripts in future. WDYT?

File: docs/plugins.md [hunk 7]
```diff
@@ -7,31 +7,36 @@ title: Plugins (Beta)
...
 ## Using Plugins
 
-Plugins are automatically loaded if you have them installed in your `package.json`. Prettier plugin package names must start with `@prettier/plugin-` or `prettier-plugin-` to be registered.
+Plugins are automatically loaded if you have them installed in the same `node_modules` directory where `prettier` is located. Plugin package names must start with `@prettier/plugin-` or `prettier-plugin-` to be registered.
 
-If the plugin is unable to be found automatically, you can load them with:
+When plugins cannot be found automatically, you can load them with:
 
-* The [CLI](./cli.md), via the `--plugin` flag:
+* The [CLI](./cli.md), via the `--plugin` and `--plugin-search-dir`:
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply docs/plugins.md]: I'm not sure why someone can't use `--plugin` and `--plugin-search-dir` simultaneously – that's technically possible and can be useful when debugging how a bunch of plugins works together (e.g. we have python and php plugins installed globally and we're testing if a locally developed elm plugin with a locally installed prettier does not conflict with the other two). Seeing both options together looks logical to me after this phrase:

> When plugins cannot be found automatically, you can load them with:
 
I don't feel there is much distinction between `--plugin` and `--plugin-search-dir` for a user, so I don't see much value in splitting the narrative into two separate sections. The only potential benefit that I see is consistency with how other CLI options are documented 🤔 

Shall we keep the docs as is until the plugin ecosystem graduates from beta and thus deserves extra attention anyway?

[the remark this answers is not in the document]
Comment (ikatyang) [reply docs/plugins.md]: I originally meant they could be used simultaneously but not required since they are completely different and won't affect each other.

After second thought, the examples are relative paths so people should understand that they're separated. This PR should be good to go now.

**Comment by kachkaev** [comment]
@ikatyang shall we merge as @duailibe suggests? I'll then rebase https://github.com/prettier/prettier/pull/4341 and add `--plugin-search-dir` to its tests.

**Comment by ikatyang** [comment]
Sure.

**Comment by azz** [comment]
:tada: Thanks for working through this problem so diligently, @kachkaev!

**Comment by suchipi** [comment]
Yes, thank you so much @kachkaev!

I'm sorry I wasn't able to give this PR the attention I should have, but I'm really glad to see this get merged. Thank you @duailibe @ikatyang @azz for reviewing in my unavailability!
